File: API/proxiesembed/media_signing.py
"""Signature HMAC des URLs média émises par le backend.

Ce module ferme la faille SSRF de `/proxy` : le service n'accepte plus une URL
arbitraire fournie par le client, mais uniquement une URL que *nous* avons
signée (soit mainapi, soit ce service lui-même en réécrivant une playlist).

Principe
--------
Chaque URL proxifiée porte deux paramètres supplémentaires :

    ?exp=<epoch_secondes>&sig=<base64url(HMAC-SHA256)>

La signature couvre `route` + `cible` + `exp`. Un client ne peut donc ni
changer la destination, ni prolonger la validité, sans connaître le secret
partagé `MEDIA_SIGNING_SECRET` (le même côté Node dans
`API/Mainapi/utils/mediaSigning.js` — les deux implémentations doivent rester
strictement alignées sur la construction du payload).

Le TTL doit couvrir la durée d'une lecture : une playlist VOD signée au début
du film sert encore ses segments à la fin. D'où un défaut généreux (12 h) —
la signature borne le rejeu, elle ne sert pas de session courte.
"""

import base64
import hashlib
import hmac
import ipaddress
import logging
import os
import re
import socket
import time
import urllib.parse
from typing import Optional, Tuple

logger = logging.getLogger(__name__)

# Secret partagé avec mainapi. Sans lui, aucune URL ne peut être signée ni
# vérifiée : le service refuse alors tout le trafic média (fail-closed).
SIGNING_SECRET = str(os.environ.get('MEDIA_SIGNING_SECRET', '') or '').strip()

# Clé interne mainapi -> proxiesembed pour les surfaces d'extraction.
INTERNAL_API_KEY = str(os.environ.get('INTERNAL_API_KEY', '') or '').strip()

INTERNAL_KEY_HEADER = 'x-internal-key'


def _get_env_int(name: str, fallback: int) -> int:
    raw = str(os.environ.get(name, '') or '').strip()
    if not raw:
        return fallback
    try:
        return int(raw)
    except (TypeError, ValueError):
        logger.warning('[signing] %s invalide, fallback=%s', name, fallback)
        return fallback


# 12 h par défaut : couvre une lecture complète, y compris une pause longue.
SIGNATURE_TTL = _get_env_int('MEDIA_SIGNATURE_TTL', 43200)

# Tolérance de dérive d'horloge entre mainapi et ce service.
CLOCK_SKEW_TOLERANCE = 300

# Paramètres réservés à la signature : ils ne doivent jamais fuiter en amont.
SIGNATURE_PARAMS = ('exp', 'sig')


class SigningNotConfigured(RuntimeError):
    """Levée quand une URL doit être signée sans secret configuré."""


def signing_configured() -> bool:
    return bool(SIGNING_SECRET)


def internal_key_configured() -> bool:
    return bool(INTERNAL_API_KEY)


def _build_payload(route: str, target: str, exp: int) -> bytes:
    # `\n` comme séparateur : il ne peut apparaître ni dans une route ni dans
    # une URL valide, donc aucune confusion de frontière n'est possible.
    return f'{route}\n{target}\n{exp}'.encode('utf-8')


def compute_signature(route: str, target: str, exp: int) -> str:
    if not SIGNING_SECRET:
        raise SigningNotConfigured('MEDIA_SIGNING_SECRET absent')
    digest = hmac.new(
        SIGNING_SECRET.encode('utf-8'),
        _build_payload(route, target, exp),
        hashlib.sha256,
    ).digest()
    return base64.urlsafe_b64encode(digest).decode('ascii').rstrip('=')


def sign_query(route: str, target: str, ttl: Optional[int] = None) -> str:
    """Retourne le fragment `exp=...&sig=...` (sans `?` ni `&` initial)."""
    exp = int(time.time()) + int(ttl if ttl is not None else SIGNATURE_TTL)
    sig = compute_signature(route, target, exp)
    return f'exp={exp}&sig={urllib.parse.quote(sig, safe="")}'


def append_signature(url: str, route: str, target: str, ttl: Optional[int] = None) -> str:
    """Ajoute la signature à une URL déjà construite (relative ou absolue)."""
    separator = '&' if '?' in url else '?'
    return f'{url}{separator}{sign_query(route, target, ttl)}'


def verify_signature(route: str, target: str, exp_raw, sig_raw) -> Tuple[bool, str]:
    """Vérifie une signature. Retourne `(valide, raison_du_refus)`."""
    if not SIGNING_SECRET:
        return False, 'signing_not_configured'
    if not exp_raw or not sig_raw:
        return False, 'missing_signature'

    try:
        exp = int(str(exp_raw).strip())
    except (TypeError, ValueError):
        return False, 'invalid_exp'

    now = int(time.time())
    if exp < now - CLOCK_SKEW_TOLERANCE:
        return False, 'expired'
    # Une expiration trop lointaine trahit une signature forgée ou un secret
    # fuité réutilisé : on borne aussi par le haut.
    if exp > now + SIGNATURE_TTL + CLOCK_SKEW_TOLERANCE:
        return False, 'exp_too_far'

    try:
        expected = compute_signature(route, target, exp)
    except SigningNotConfigured:
        return False, 'signing_not_configured'

    # compare_digest : temps constant, pas d'oracle de timing sur la signature.
    if not hmac.compare_digest(expected, str(sig_raw).strip()):
        return False, 'bad_signature'

    return True, ''


def verify_request(request, route: str, target: str) -> Tuple[bool, str]:
    """Vérifie la signature portée par les query params d'une requête aiohttp."""
    return verify_signature(
        route,
        target,
        request.query.get('exp'),
        request.query.get('sig'),
    )


def check_internal_key(request) -> bool:
    """Vrai si la requête porte la clé interne mainapi -> proxiesembed."""
    if not INTERNAL_API_KEY:
        return False
    provided = request.headers.get(INTERNAL_KEY_HEADER, '')
    if not provided:
        return False
    return hmac.compare_digest(INTERNAL_API_KEY, provided.strip())


def redact_url(url: Optional[str]) -> str:
    """Réduit une URL à `schéma://hôte/chemin` pour les logs.

    Les query strings des CDN portent des jetons d'authentification : les
    journaliser telles quelles les expose à quiconque lit les logs.
    """
    if not url:
        return '(none)'
    try:
        parsed = urllib.parse.urlparse(str(url))
        if not parsed.netloc:
            return '(invalid-url)'
        suffix = '?…' if parsed.query else ''
        return f'{parsed.scheme}://{parsed.netloc}{parsed.path}{suffix}'
    except (TypeError, ValueError):
        return '(invalid-url)'


# ---------------------------------------------------------------------------
#  Encodage base signée pour les BaseURL DASH (/drm/b/<blob>/<subpath>)
# ---------------------------------------------------------------------------
# Une BaseURL DASH est un préfixe de chemin : y accrocher `?exp=&sig=` casserait
# la résolution des URLs relatives que le player construit par-dessus. On glisse
# donc la signature dans le blob base64 lui-même.

DRM_BASE_ROUTE = '/drm/b'


def encode_signed_drm_base(base_url: str, ttl: Optional[int] = None) -> str:
    exp = int(time.time()) + int(ttl if ttl is not None else SIGNATURE_TTL)
    sig = compute_signature(DRM_BASE_ROUTE, base_url, exp)
    blob = f'{exp}:{sig}:{base_url}'
    return base64.urlsafe_b64encode(blob.encode('utf-8')).decode('ascii').rstrip('=')


def decode_signed_drm_base(blob_b64: str) -> Tuple[Optional[str], str]:
    """Décode et vérifie un blob de BaseURL. Retourne `(base_url, raison)`."""
    padding = (4 - len(blob_b64) % 4) % 4
    try:
        decoded = base64.urlsafe_b64decode(blob_b64 + '=' * padding).decode('utf-8')
    except Exception:
        return None, 'invalid_encoding'

    parts = decoded.split(':', 2)
    if len(parts) != 3:
        return None, 'invalid_payload'

    exp_raw, sig_raw, base_url = parts
    valid, reason = verify_signature(DRM_BASE_ROUTE, base_url, exp_raw, sig_raw)
    if not valid:
        return None, reason
    return base_url, ''


# ---------------------------------------------------------------------------
#  Garde anti-SSRF sur la destination
# ---------------------------------------------------------------------------
# Défense en profondeur : même signée, une URL ne doit jamais viser le réseau
# interne. Protège contre un secret fuité et contre nos propres bugs de
# construction d'URL.

_BLOCKED_HOSTNAMES = frozenset({
    'localhost',
    'localhost.localdomain',
    'ip6-localhost',
    'ip6-loopback',
    # Endpoints de métadonnées cloud : cible SSRF classique (vol de creds IAM).
    'metadata',
    'metadata.google.internal',
    'metadata.goog',
})


def is_public_http_url(url: str) -> bool:
    """Vrai si l'URL est http(s) et ne vise pas une adresse interne.

    N'effectue pas de résolution DNS : ce contrôle porte sur ce qui est
    littéralement écrit dans l'URL. La protection contre le DNS rebinding
    relève du resolver (cf. `PublicOnlyResolver`).
    """
    if not url or not isinstance(url, str):
        return False

    try:
        parsed = urllib.parse.urlparse(url)
    except (TypeError, ValueError):
        return False

    if parsed.scheme.lower() not in ('http', 'https'):
        return False

    hostname = (parsed.hostname or '').strip().rstrip('.').lower()
    if not hostname:
        return False

    if hostname in _BLOCKED_HOSTNAMES:
        return False

    # `.internal`, `.local`, `.localhost` : espaces de noms non routables.
    if hostname.endswith(('.internal', '.local', '.localhost', '.home.arpa')):
        return False

    ip = _parse_ip_literal(hostname.strip('[]'))
    if ip is None:
        # Nom de domaine : on ne peut rien conclure sans résolution DNS.
        return True

    # Une adresse IPv4 encapsulée dans de l'IPv6 doit être jugée sur l'IPv4
    # qu'elle transporte, sinon `::ffff:127.0.0.1` passerait pour publique.
    mapped = getattr(ip, 'ipv4_mapped', None)
    if mapped is not None:
        ip = mapped

    return not (
        ip.is_private
        or ip.is_loopback
        or ip.is_link_local
        or ip.is_multicast
        or ip.is_reserved
        or ip.is_unspecified
    )


def _parse_ip_literal(host: str):
    """Reconnaît une IP écrite sous n'importe quelle forme acceptée par la résolution.

    `ipaddress` ne connaît que la notation canonique, alors que la résolution
    système accepte aussi les formes entière (`2130706433`), octale
    (`0177.0.0.1`) et abrégée (`127.1`) — toutes équivalentes à 127.0.0.1.
    Sans cette normalisation, elles étaient prises pour des noms de domaine et
    échappaient au contrôle.

    Retourne None si `host` n'est pas une IP littérale.
    """
    try:
        return ipaddress.ip_address(host)
    except ValueError:
        pass

    # `inet_aton` couvre exactement les formes tolérées par le résolveur.
    try:
        packed = socket.inet_aton(host)
    except (OSError, UnicodeEncodeError):
        return None

    # inet_aton accepte aussi des noms purement numériques déjà gérés au-dessus ;
    # ici on ne retient que ce qui n'est pas un nom de domaine plausible.
    if not re.fullmatch(r'[0-9a-fA-FxX.]+', host):
        return None

    try:
        return ipaddress.IPv4Address(packed)
    except ValueError:
        return None

File: API/Mainapi/utils/mediaSigning.js
/**
 * mediaSigning.js — Signature HMAC des URLs média servies par proxiesembed.
 *
 * Pendant Node de `API/proxiesembed/media_signing.py`. Les deux
 * implémentations DOIVENT construire le payload à l'identique :
 *
 *     payload = `${route}\n${target}\n${exp}`
 *     sig     = base64url(HMAC-SHA256(MEDIA_SIGNING_SECRET, payload))  // sans '='
 *
 * Toute divergence (ordre des champs, séparateur, padding base64) casse
 * silencieusement la lecture vidéo côté client : ne modifier qu'en binôme.
 */

const crypto = require('crypto');
const net = require('net');

const SIGNING_SECRET = (process.env.MEDIA_SIGNING_SECRET || '').trim();
const INTERNAL_API_KEY = (process.env.INTERNAL_API_KEY || '').trim();

function parseIntEnv(name, fallback) {
    const raw = (process.env[name] || '').trim();
    if (!raw) return fallback;
    const parsed = Number.parseInt(raw, 10);
    return Number.isFinite(parsed) ? parsed : fallback;
}

// 12 h — doit rester identique au défaut Python (MEDIA_SIGNATURE_TTL).
const SIGNATURE_TTL = parseIntEnv('MEDIA_SIGNATURE_TTL', 43200);

const INTERNAL_KEY_HEADER = 'x-internal-key';

function signingConfigured() {
    return SIGNING_SECRET.length > 0;
}

function internalKeyConfigured() {
    return INTERNAL_API_KEY.length > 0;
}

function computeSignature(route, target, exp) {
    if (!SIGNING_SECRET) {
        throw new Error('MEDIA_SIGNING_SECRET absent');
    }
    return crypto
        .createHmac('sha256', SIGNING_SECRET)
        .update(`${route}\n${target}\n${exp}`, 'utf8')
        .digest('base64url');
}

/** Fragment `exp=...&sig=...`, sans `?` ni `&` initial. */
function signQuery(route, target, ttlSeconds) {
    const ttl = Number.isFinite(ttlSeconds) ? ttlSeconds : SIGNATURE_TTL;
    const exp = Math.floor(Date.now() / 1000) + ttl;
    const sig = computeSignature(route, target, exp);
    return `exp=${exp}&sig=${encodeURIComponent(sig)}`;
}

/** Ajoute la signature à une URL déjà construite. */
function appendSignature(url, route, target, ttlSeconds) {
    const separator = url.includes('?') ? '&' : '?';
    return `${url}${separator}${signQuery(route, target, ttlSeconds)}`;
}

/**
 * Construit une URL proxifiée signée, prête à être consommée par le player.
 *
 * @param {string} proxyBase  Origine publique de proxiesembed (sans slash final)
 * @param {string} route      Route proxy, ex. '/proxy' ou '/cinep-proxy'
 * @param {string} target     URL amont à relayer
 * @param {object} [options]  { ttlSeconds, extraParams }
 */
function buildSignedProxyUrl(proxyBase, route, target, options = {}) {
    const base = String(proxyBase || '').replace(/\/+$/, '');
    const params = new URLSearchParams({ url: target });
    for (const [key, value] of Object.entries(options.extraParams || {})) {
        if (value !== undefined && value !== null && value !== '') {
            params.set(key, String(value));
        }
    }
    const url = `${base}${route}?${params.toString()}`;
    return appendSignature(url, route, target, options.ttlSeconds);
}

// ---------------------------------------------------------------------------
//  Jetons opaques signés
// ---------------------------------------------------------------------------
// Quand une valeur d'origine serveur doit faire un aller-retour par le client
// (l'adresse d'une page France.tv, par exemple), on ne la lui confie pas en
// clair : il pourrait la remplacer. On l'emballe dans un jeton signé, que lui
// seul transporte sans pouvoir le fabriquer.
//
// Même construction que `encode_signed_drm_base` côté Python : `exp:sig:valeur`
// en base64url.

function encodeSignedToken(route, value, ttlSeconds) {
    const ttl = Number.isFinite(ttlSeconds) ? ttlSeconds : SIGNATURE_TTL;
    const exp = Math.floor(Date.now() / 1000) + ttl;
    const sig = computeSignature(route, value, exp);
    return Buffer.from(`${exp}:${sig}:${value}`, 'utf8').toString('base64url');
}

/**
 * Décode et vérifie un jeton. Retourne la valeur d'origine, ou null si le
 * jeton est absent, malformé, expiré ou forgé.
 */
function decodeSignedToken(route, token) {
    if (!token || typeof token !== 'string' || !SIGNING_SECRET) return null;

    let decoded;
    try {
        decoded = Buffer.from(token, 'base64url').toString('utf8');
    } catch {
        return null;
    }

    // La valeur peut contenir des « : » (une URL en contient) : on ne découpe
    // que sur les deux premiers.
    const first = decoded.indexOf(':');
    const second = decoded.indexOf(':', first + 1);
    if (first < 0 || second < 0) return null;

    const exp = Number.parseInt(decoded.slice(0, first), 10);
    const sig = decoded.slice(first + 1, second);
    const value = decoded.slice(second + 1);
    if (!Number.isInteger(exp) || !sig || !value) return null;

    const now = Math.floor(Date.now() / 1000);
    if (exp < now || exp > now + SIGNATURE_TTL + 300) return null;

    let expected;
    try {
        expected = computeSignature(route, value, exp);
    } catch {
        return null;
    }

    const a = Buffer.from(expected);
    const b = Buffer.from(sig);
    if (a.length !== b.length || !crypto.timingSafeEqual(a, b)) return null;

    return value;
}

/** Headers à joindre à un appel sortant mainapi -> proxiesembed. */
function internalHeaders(extra = {}) {
    if (!INTERNAL_API_KEY) {
        throw new Error('INTERNAL_API_KEY absent');
    }
    return { ...extra, [INTERNAL_KEY_HEADER]: INTERNAL_API_KEY };
}

// ---------------------------------------------------------------------------
//  Garde anti-SSRF sur la destination
// ---------------------------------------------------------------------------
// mainapi signe des URLs à la demande (Live TV). Sans ce contrôle, un porteur
// de clé VIP transformerait l'endpoint de signature en oracle SSRF capable de
// viser le réseau interne.

const BLOCKED_HOSTNAMES = new Set([
    'localhost',
    'localhost.localdomain',
    'ip6-localhost',
    'ip6-loopback',
    'metadata',
    'metadata.google.internal',
    'metadata.goog',
]);

const BLOCKED_SUFFIXES = ['.internal', '.local', '.localhost', '.home.arpa'];

function isPrivateIpv4(address) {
    const octets = address.split('.').map(Number);
    if (octets.length !== 4 || octets.some((o) => !Number.isInteger(o) || o < 0 || o > 255)) {
        return true;
    }
    const [a, b] = octets;
    if (a === 0 || a === 10 || a === 127) return true;
    if (a === 169 && b === 254) return true;            // link-local + métadonnées cloud
    if (a === 172 && b >= 16 && b <= 31) return true;
    if (a === 192 && b === 168) return true;
    if (a === 100 && b >= 64 && b <= 127) return true;  // CGNAT
    if (a === 192 && b === 0) return true;              // IETF protocol assignments
    if (a >= 224) return true;                          // multicast + réservé
    return false;
}

function isPrivateIpv6(address) {
    const normalized = address.toLowerCase().replace(/^\[|\]$/g, '');
    if (normalized === '::' || normalized === '::1') return true;
    if (normalized.startsWith('fe80')) return true;                 // link-local
    if (/^f[cd]/.test(normalized)) return true;                     // unique-local

    // IPv4-mapped (::ffff:a.b.c.d). Attention : le parseur d'URL de Node
    // recompresse la forme pointée en hexadécimal — `[::ffff:127.0.0.1]`
    // devient `::ffff:7f00:1`. Ne comparer que la forme pointée laissait donc
    // passer le loopback. On traite les deux écritures.
    if (normalized.startsWith('::ffff:')) {
        const mapped = normalized.slice('::ffff:'.length);
        if (net.isIPv4(mapped)) return isPrivateIpv4(mapped);
        const hextets = mapped.split(':');
        if (hextets.length === 2) {
            const high = Number.parseInt(hextets[0], 16);
            const low = Number.parseInt(hextets[1], 16);
            if (Number.isInteger(high) && Number.isInteger(low)) {
                const dotted = [high >> 8, high & 0xff, low >> 8, low & 0xff].join('.');
                return isPrivateIpv4(dotted);
            }
        }
        // Forme mappée non reconnue : on refuse par défaut.
        return true;
    }
    return false;
}

/**
 * Vrai si l'URL est http(s) et ne vise pas littéralement une adresse interne.
 * Ne résout pas le DNS — la protection contre le rebinding est côté resolver.
 */
function isPublicHttpUrl(value) {
    if (!value || typeof value !== 'string') return false;

    let parsed;
    try {
        parsed = new URL(value);
    } catch {
        return false;
    }

    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') return false;

    const hostname = parsed.hostname.trim().replace(/\.$/, '').toLowerCase();
    if (!hostname) return false;
    if (BLOCKED_HOSTNAMES.has(hostname)) return false;
    if (BLOCKED_SUFFIXES.some((suffix) => hostname.endsWith(suffix))) return false;

    if (net.isIPv4(hostname)) return !isPrivateIpv4(hostname);

    const unbracketed = hostname.replace(/^\[|\]$/g, '');
    if (net.isIPv6(unbracketed)) return !isPrivateIpv6(unbracketed);

    return true;
}

module.exports = {
    INTERNAL_KEY_HEADER,
    SIGNATURE_TTL,
    signingConfigured,
    internalKeyConfigured,
    computeSignature,
    signQuery,
    appendSignature,
    buildSignedProxyUrl,
    encodeSignedToken,
    decodeSignedToken,
    internalHeaders,
    isPublicHttpUrl,
};
